### PR TITLE
[fix](string64) fix coredump caused by ColumnArray<ColumnStr<uint64_t>>::insert_indices_from

### PR DESCRIPTION
**Pull request by jacktengg**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:
In order to avoid the problem of  total string length of a `ColumnStr<uint32_t>` column exceeds the 4G limit, hash join operator call `MutableBlock::merge_ignore_overflow` to collect build data.

After finishing collecting build data, a `ColumnStr<uint32_t>` column will be converted to `ColumnStr<uint64_t>` if the total string length exceeds the 4G limit by calling `Block::replace_if_overflow`.

When hash join operator ouput join result block, it need to calls `ColumnStr<T>::insert_indices_from` to fill  a string column, which is able to handle both `ColumnStr<uint32_t>` and `ColumnStr<uint64_t>`. Every column which may include string sub-columns should also call `insert_indices_from` method of it's sub-column to handle both `ColumnStr<uint32_t>` and `ColumnStr<uint64_t>`, or else column cast will fail.

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [x] No.
    - [ ] Yes. 

- Does this need documentation?
    - [x] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run buildall

**Comment by jacktengg** [comment]
run buildall

File: be/src/vec/columns/column_string.cpp [hunk 1]
```diff
@@ -134,34 +133,41 @@ void ColumnStr<T>::insert_range_from(const IColumn& src, size_t start, size_t le
...
-                    src_concrete.offsets[start + i] - nested_offset + prev_max_offset;
+            for (size_t i = 0; i < length; ++i) {
+                offsets[old_size + i] = src_offsets[start + i] - nested_offset + prev_max_offset;
+            }
         }
+    };
+    // insert_range_from maybe called by ColumnArray::insert_indices_from(which is used by hash join operator),
+    // so we need to support both ColumnStr<uint32_t> and ColumnStr<uint64_t>
+    if (src.is_column_string64()) {
+        do_insert(assert_cast<const ColumnStr<uint64_t>&>(src));
```
Comment (yiguolei) [inline be/src/vec/columns/column_string.cpp:168]: If src is string64 but dest is string32？

File: be/src/vec/columns/column_string.cpp [hunk 2]
```diff
@@ -134,34 +133,41 @@ void ColumnStr<T>::insert_range_from(const IColumn& src, size_t start, size_t le
     if (length == 0) {
         return;
     }
+    auto do_insert = [&](const auto& src_concrete) {
```
Comment (yiguolei) [inline be/src/vec/columns/column_string.cpp:136]: clone_resized  method should check if to size is larger than limit

File: be/src/vec/columns/column_string.cpp [hunk 3]
```diff
@@ -134,34 +133,41 @@ void ColumnStr<T>::insert_range_from(const IColumn& src, size_t start, size_t le
...
+        const auto& src_offsets = src_concrete.get_offsets();
+        const auto& src_chars = src_concrete.get_chars();
```
Comment (yiguolei) [inline be/src/vec/columns/column_string.cpp:138]: add comment to insert_range_from_ignore_overflow

File: be/src/vec/columns/column_string.cpp [hunk 4]
```diff
@@ -134,34 +133,41 @@ void ColumnStr<T>::insert_range_from(const IColumn& src, size_t start, size_t le
...
 
-    if (start == 0 && offsets.empty()) {
-        offsets.assign(src_concrete.offsets.begin(), src_concrete.offsets.begin() + length);
-    } else {
-        size_t old_size = offsets.size();
-        size_t prev_max_offset = offsets.back(); /// -1th index is Ok, see PaddedPODArray
-        offsets.resize(old_size + length);
+        using OffsetsType = std::decay_t<decltype(src_offsets)>;
+        if (std::is_same_v<T, typename OffsetsType::value_type> && start == 0 && offsets.empty()) {
+            offsets.assign(src_offsets.begin(), src_offsets.begin() + length);
```
Comment (yiguolei) [inline be/src/vec/columns/column_string.cpp:154]: 这种if 确实有用么。。。 我感觉161 行那个写法，编译器应该也会循环展开的，不一定慢啊

File: be/src/vec/columns/column_string.cpp (hunk 1 above)
Comment (jacktengg) [reply be/src/vec/columns/column_string.cpp:168]: offset64 will be converted to offset32, and check_chars_length is called to check if total len exceed 4G

File: be/src/vec/columns/column_string.cpp (hunk 2 above)
Comment (jacktengg) [reply be/src/vec/columns/column_string.cpp:136]: if `to_size` is bigger than `from_size`, extra element is empty string, total stringg size will not bigger than src size.

**Comment by jacktengg** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**